### PR TITLE
RPCh integration 

### PR DESCRIPTION
**Pull request by kjesien**
## Scope

The purpose of this PR is to enable RPCh integration for the Enkrypt browser extension.

[Issue](https://github.com/Rpc-h/extension-enKrypt/issues/1)

## Implementation

Enkrypt extension consists of a few modules, the most important are:
- `extension`  - actual browser extension providing crypto wallet features
- `request` - a module used to provide the provider which is injected to the `window` object of visited websites allowing interaction with dApps

Both use separate mechanisms connecting to the RPCs.
Important note it that currently, only HTTP-based requests are available - not Web Socket-based.

### `extension`
Communication is based on the `Eth` class provided by the `web3-eth` package which is the official package of web3.js.
It provides a few options for supplying different kinds of provider objects like HttpProvider, WebsocketProvider and AbstractProvider. 
I decided to implement the class using the AbstractProvider interface, having the method - `sendAsync` which uses the RPCh SDK instance to send the requests.
~~I decided to use the singleton pattern for the SDK instance for the sake of optimization in contrast to having a separate RPCh SDK instance for every provider.~~ After code review from @esterlus I changed to use separate SDK instances for each provider.
Config for the SDK is passed using .env file. Exemplary file provided.

### `request`
Previously it used `JSONRPCClient` from `json-rpc-2.0` with the `request` method so I decided to implement a separate class mimicking it, but using RPCh.
Config for the SDK is passed using the CLI parameter, for details go to the readme file.

### Enkrypt UI

[demo.webm](https://github.com/Rpc-h/extension-enKrypt/assets/9512544/8cc15c01-104c-49b2-bec2-883493457760)

### Enkrypt UI

1. Clone/check out the branch `1-rpch-sdk-integration-b2b-use-case`
2. `yarn install`
3. `yarn build:all`
4. `yarn build:chrome`
6.  Add to chrome unpacked extension - [how to](https://developer.chrome.com/docs/extensions/mv2/getstarted/#manifest) using directory `packages\extension\dist`
7. All network RPCs preselected by Enkrypt are web socket-based so please add manually network using HTTP RPC for example `https://derp.hoprnet.org/rpc/eth/mainnet` or for Sepolia `https://rpc2.sepolia.org/`

### Request module

~~Unfortunately, I was not able to test it as I don't use any dApps, but I assume that it will just by issuing transactions from the dApp itself.~~
1. In the extension popup select network using HTTP RPC Sepolia
2. Go to any website
3. Open Dev tools
4. Use the following snippet to fetch the balance of the account:
```
await enkrypt.providers.ethereum.sendMessageHandler(
        'ethereum',
        JSON.stringify({
          method: "eth_getBalance",
          params: ["0x00000000219ab540356cBB839Cbe05303d7705Fa", "latest"],
        })
      )
```

### CODE REVIEW

File: packages/request/tsconfig.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "noUnusedParameters": true,
     "removeComments": true,
     "sourceMap": true,
-    "target": "es5",
+    "target": "es6",
```
Comment (esterlus) [inline packages/request/tsconfig.json:10]: is this needed?

File: packages/request/tsup.config.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { defineConfig } from "tsup";
+import { config } from "dotenv";
+
+const env = {};
+config({ processEnv: env });
+
+// TODO: Remove after confirmation and testing
+console.log("REQUEST PACKAGE ENV VARS", env);
```
Comment (esterlus) [inline packages/request/tsup.config.ts:8]: why add a whole config file for tsup here?

File: packages/request/package.json [hunk 3]
```diff
@@ -32,6 +33,7 @@
     "@typescript-eslint/eslint-plugin": "^5.62.0",
     "@typescript-eslint/parser": "^5.62.0",
     "chai": "^4.3.8",
+    "dotenv": "^16.3.1",
```
Comment (esterlus) [inline packages/request/package.json:36]: Not a good idea to add a new dev dependency

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      .then(async (res) => {
+        const jsonRes: Result | sdkError = await res.json();
+        if (this.isError(jsonRes)) {
+          callback?.({ name: "", ...jsonRes.error });
+          return;
+        }
+
+        const parsedRes = {
+          ...jsonRes,
+          id: +(jsonRes.id || 0),
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:70]: why do you override the id of the response?

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+// TODO: Remove after confirmation and testing
+console.log("RPCh: CREATING SDK INSTANCE with OPS ", ops);
+console.log("RPCh: Client ID ", process.env.VUE_APP_RPCH_SECRET_TOKEN);
+
+if (!process.env.VUE_APP_RPCH_SECRET_TOKEN) {
+  throw "MISSING RPCH SECRET TOKEN";
+}
+
+const RPChSdk = new RPChSDK(process.env.VUE_APP_RPCH_SECRET_TOKEN, ops);
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:19]: should be a member of the class here

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+export const getSuportedRpchProvider = (
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:21]: typo

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+export class RPChProvider implements AbstractProvider {
+  rpcUrl: string;
+
+  constructor(rpcUrl: string) {
+    this.rpcUrl = rpcUrl;
+  }
+
+  private isError(jsonRes: Result | sdkError): jsonRes is sdkError {
+    return Boolean("error" in jsonRes);
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:38]: might be a redundant Boolean conversion ?? wouldn't `return "error" in jsonRes;` be the same?

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      " params: ",
+      payload.params
+    );
+
+    RPChSdk.send(
+      {
+        ...payload,
+        jsonrpc: "2.0",
+      },
+      { provider: this.rpcUrl }
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:59]: once sdk is member of the class you don't need to override the provider on every request

File: packages/extension/src/libs/hoprProvider/index.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    )
+      .then(async (res) => {
+        const jsonRes: Result | sdkError = await res.json();
+        if (this.isError(jsonRes)) {
+          callback?.({ name: "", ...jsonRes.error });
```
Comment (esterlus) [inline packages/extension/src/libs/hoprProvider/index.ts:64]: don't know if you need to handle it this way but generally speaking a jsonrpcerror is not the same as a HTTP req error or connection error.

However here they are treated the same

File: packages/extension/src/providers/ethereum/ui/send-transaction/index.vue [hunk 10]
```diff
@@ -356,7 +357,7 @@ const setTransactionFees = (tx: Transaction) => {
 };
 
 const setBaseCosts = () => {
-  const web3 = new Web3Eth(props.network.node);
+  const web3 = new Web3Eth(new HoprProvider(props.network.node) as any);
```
Comment (esterlus) [inline packages/extension/src/providers/ethereum/ui/send-transaction/index.vue:360]: HoprProvider??

**Review by esterlus — COMMENTED** [review]
looks good, just some minor improvement requests

File: packages/request/tsup.config.ts (hunk 2 above)
Comment (kjesien) [reply packages/request/tsup.config.ts:8]: I had to add it to be able to use environmental variables.

File: packages/request/package.json (hunk 3 above)
Comment (kjesien) [reply packages/request/package.json:36]: I had to add it to be able to use environmental variables.

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 4 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: It's for the sake of types compliance between `Result` from `@rpch/sdk` and `JsonRpcResponse` from `web3-core`

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 5 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: It's a question I wanted to raise. 
For every use of `Web3Eth`, a new instance of `RPChProvider` is required, so if an SDK instance would be part of it, we would end up with multiple SDK instances.
I was concerned about performance and network traffic, so I decided to use SDK as a singleton - so having a single instance of SDK across the App.
What do you think about it?

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 6 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: Thanks

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 9 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: It's a matter of type compliance.

File: packages/extension/src/providers/ethereum/ui/send-transaction/index.vue (hunk 10 above)
Comment (kjesien) [reply packages/extension/src/providers/ethereum/ui/send-transaction/index.vue]: Project does not have proper linting or even build - it was not detected by processes, but fixed. :)

File: packages/request/tsconfig.json (hunk 1 above)
Comment (kjesien) [reply packages/request/tsconfig.json:10]: Unfortunately yes. To use environment variables I had to use tsup config file but when I added it I got build errors like:
```
ERROR: Transforming const to the configured target environment ("es5") is not supported yet
```
So I had to bump the target.

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 8 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: It's connected to the problems raised here - https://github.com/Rpc-h/extension-enKrypt/pull/2#discussion_r1383641055

File: packages/request/tsconfig.json (hunk 1 above)
Comment (kjesien) [reply packages/request/tsconfig.json]: Reverted.

File: packages/request/package.json (hunk 3 above)
Comment (kjesien) [reply packages/request/package.json]: Reverted.

File: packages/extension/src/libs/hoprProvider/index.ts (hunk 5 above)
Comment (kjesien) [reply packages/extension/src/libs/hoprProvider/index.ts]: Resolved

File: packages/extension/src/libs/RPChProvider/index.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  ) {
+    this.sdk
+      .send({
+        ...payload,
+        jsonrpc: "2.0",
+      })
+      .then(async (res) => {
+        const jsonRes: Result | sdkError = await res.json();
+        if (this.isError(jsonRes)) {
+          callback?.({ name: "", ...jsonRes.error });
```
Comment (esterlus) [inline packages/extension/src/libs/RPChProvider/index.ts:51]: This is not an error. A valid response can have "error" as an attribtue

File: .gitignore [hunk 12]
```diff
@@ -116,3 +116,5 @@ dist
 .DS_Store
 
 .nova/
+
+.idea
```
Comment (esterlus) [inline .gitignore:120]: please restore file to original

File: packages/extension/.env.example [hunk 13]
```diff
@@ -0,0 +1,2 @@
+VUE_APP_DISCOVERY_PLATFORM_API_ENDPOINT=
```
Comment (esterlus) [inline packages/extension/.env.example:1]: please remove

File: packages/request/.env.example [hunk 14]
```diff
@@ -0,0 +1,2 @@
+DISCOVERY_PLATFORM_API_ENDPOINT=
```
Comment (esterlus) [inline packages/request/.env.example:1]: please remove

**Review by esterlus — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 12 above)
Comment (kjesien) [reply .gitignore]: Removed

File: packages/extension/.env.example (hunk 13 above)
Comment (kjesien) [reply packages/extension/.env.example:1]: Removed

File: packages/request/.env.example (hunk 14 above)
Comment (kjesien) [reply packages/request/.env.example]: Removed

File: packages/extension/src/libs/RPChProvider/index.ts (hunk 11 above)
Comment (kjesien) [reply packages/extension/src/libs/RPChProvider/index.ts]: Done.

**Review by esterlus — APPROVED** [review]

**Pull request merged**